[Woo POS] M2 Error: Checking order

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13479
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Handle 'Checking order' stage errors.

<img width="448" alt="image" src="https://github.com/user-attachments/assets/441b3d51-d707-48ca-868f-8b9564499532">

⚠️  There're 2 differences from the designs:
1. No "Learn More" button. There's no further information to display when this error happens
2. No darker background. We should do this in a separate task since it affects other views as well (https://github.com/woocommerce/woocommerce-ios/issues/13549). We cannot just set a background color for message view, since `TotalsView` has its centering, spacing, and padding. We likely need to restructure `TotalsView` so cardReaderView would be placed inside a larger container, which would touch the leading, trailing, and top sides. When required, we can change the color of this container, otherwise cardReaderView placement would be unaffected.

## Solution

Error when validating order (`CollectOrderPaymentUseCaseError.couldNotRefreshOrde`r) needs to be presented separately, therefore:

1. created a new `validatingOrderErro`r message type
2 Mapped `CollectOrderPaymentUseCaseError.couldNotRefreshOrder` `paymentError` to `validatingOrderError` message type
3. Created a new message view sharing most of layout configuration with ReaderDisconnectedMessageView

## Testing information


1. Either use CharlesProxy/Proxyman to trigger an error, or make changes to `OrderState retrieveOrderRemotely` to pass a wrong `siteId` can trigger one of the errors.
2. Open POS
2. Connect Reader
3. Add Items to Cart
4. Check Out
5. Notice error appears after validating order

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/304b1afe-0dba-4915-b693-26e583dc6e2a

---
- [x] I have considered adding unit tests for this change. If I decided not to add them, I have provided a brief explanation below (optional):
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [x] This PR includes refactoring; smoke testing of the entire section is needed.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Create validatingOrderError for couldNotRefreshOrder paymentError

Error when validating order (CollectOrderPaymentUseCaseError.couldNotRefreshOrder) needs to be presented separately, therefore:

1. created a new validatingOrderError message type
1. Map CollectOrderPaymentUseCaseError.couldNotRefreshOrder paymentError to validatingOrderError message type

staskus committed:
Create PointOfSaleCardPresentPaymentValidatingOrderErrorMessageView to display validatingOrderError

staskus committed:
Merge branch 'trunk' into task/13479-pos-m2-error-checking-order

staskus committed:
Use WooCommerceAmber from ColorPalette

staskus committed:
Use POSErrorExclamationMark() in all message views

staskus committed:
Print specific DotcomError when order validation fails

staskus committed:
Use consistent padding with other error views

staskus commented on WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift:
```diff
@@ -127,9 +127,10 @@ extension CardPresentPaymentEventDetails {
-
         case .paymentSuccess:
             return .message(.paymentSuccess(viewModel: PointOfSaleCardPresentPaymentSuccessMessageViewModel()))
+        case .paymentError(error: CollectOrderPaymentUseCaseError.couldNotRefreshOrder(let error), let tryAgain, _):
```
@joshheald, this is where I need your opinion.

For now, order validation is one of the payment errors. Theoretically, I could introduce `validatingOrderError` at any stage. 

I made a decision to only introduce `validatingOrderError` when translating `CardPresentPaymentEvent` into `PointOfSaleCardPresentPaymentEventPresentationStyle`. However, you may have a more informed view on this.

Thanks!

joshheald replied:
I agree with this approach

This order validation is part of the payment process, and it's in response to a `CardPresentPaymentEvent` produced as one of the first steps after starting a payment.

This happens after refreshing the order to check it's still valid for payment. IMO the POS doesn't really need to do that, because the order is only just created and won't be accessible later, so should be valid. However, it's part of the underlying payment code that we reuse, so it's better to handle it than try to prevent it from happening.

We may in future have other order validation errors, when they're added they should be handled elsewhere.

joshheald reviewed (approved):
Works as expected when the request simply fails.

There are some other cases which I think should be treated in the same way.

[See this (underlying) payment code for the possible errors when validating the order](https://github.com/woocommerce/woocommerce-ios/blob/535f41d2a519bac8aab541b1c22952229e6fe041/WooCommerce/Classes/ViewRelated/Orders/Collect%20Payments/CollectOrderPaymentUseCase.swift#L250-L271)

`orderAlreadyPaid` will be difficult to reproduce there, and I think you can ignore it – [we already specifically adapt it to show it as `success`](https://github.com/woocommerce/woocommerce-ios/blob/535f41d2a519bac8aab541b1c22952229e6fe041/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentsAlertPresenterAdaptor.swift#L19-L21). It will usually happen when you collect a payment successfully, get another error, then retry – we don't want people to double charge in that case.

I think if you add some handling for `totalAmountInvalid` and `orderTotalChanged`, this'll be a bit more comprehensive.

We could make the card payment adaptor produce a `validatingOrderFailed` event here if that would be helpful, rather than have them as generic payment events. Let me know if you want to go through that.

staskus commented:
> There are some other cases which I think should be treated in the same way.

Thanks @joshheald  I definitely missed that 👍 

> I think if you add some handling for totalAmountInvalid and orderTotalChanged, this'll be a bit more comprehensive.

Sounds reasonable 👍 

> We could make the card payment adaptor produce a validatingOrderFailed event here if that would be helpful, rather than have them as generic payment events. Let me know if you want to go through that.

Do you think it's a good option? 🤔 I'll check what seems to make the most sense to handle multiple error types.

joshheald commented:
> > We could make the card payment adaptor produce a validatingOrderFailed event here if that would be helpful, rather than have them as generic payment events. Let me know if you want to go through that.
> 
> Do you think it's a good option? 🤔 I'll check what seems to make the most sense to handle multiple error types.

Perhaps, I'm not 100% on it without trying. 

While you're doing it, it's important to think about the adaptor code and anything with a `CardPresentPayment` prefix as being generic and reusable in other UI cases, while the code with `PointOfSale` prefixes can be specific to POS.

For example, I'm changing `CardPresentPaymentEventDetails.paymentError` and `.paymentErrorNonRetryable` to just be `.paymentError`, but take a `retryApproach` enum. That's something that's probably needed by any caller of the `CardPresentPaymentService`, so that they can show the correct UI for different types of retry (same card, different card, no retry.)

I think it could make sense to have `validateOrderFailed(reason:)` as a `CardPresentPaymentEventDetails`, as other callers may also want to show all these order validation errors with a common style, separate from errors that happen after the card is tapped. However, it might just be making it harder to show errors for future callers of the service.

staskus committed:
Handle additional errors as other validatingOrderErrors

staskus committed:
Update PointOfSaleCardPresentPaymentEventPresentationStyle.swift

staskus commented on WooCommerce/Classes/POS/Presentation/Card Present Payments/PointOfSaleCardPresentPaymentEventPresentationStyle.swift:
```diff
@@ -127,20 +127,29 @@ extension CardPresentPaymentEventDetails {
-                viewModel: PointOfSaleCardPresentPaymentNonRetryableErrorMessageViewModel(
-                    error: error)))
-
+            switch error {
```
@joshheald added handling of additional order validation-related errors. Those are all non-retryable so I'm hiding a retry button:

https://github.com/user-attachments/assets/002a8a80-eadd-4916-bc80-a81444a5223a

staskus committed:
Merge branch 'trunk' into task/13479-pos-m2-error-checking-order

staskus committed:
Add validatingOrderError paymentState to configure TotalsView

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/218">20.0</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
